orchestrator: Minimize blockchain lookups

j0sh opened the pull request:
Fixes https://github.com/livepeer/go-livepeer/issues/517

* [x] orch.CurrentBlock
* [x] DB query to fetch a specific job
* [x] orch.GetJob
* [x] Add tests for all the above
* [x] Claimmanager GetJob to use DB

j0sh committed:
orchestrator: Pull current block from DB.

j0sh committed:
orchestrator: Pull latest job from DB when possible.

j0sh committed:
core: Clean up leftover generated files from tests.

darkdarkdragon commented on common/db.go:
```diff
@@ -526,6 +539,32 @@ func (db *DB) ActiveJobs(since *big.Int) ([]*DBJob, error) {
+	profiles, err := TxDataToVideoProfile(options)
+	if err != nil {
+		glog.Error("Unable to convert transcode options into ffmpeg profile ", err)
+	}
```
Right now `TxDataToVideoProfile` never returns any error, but potentially can return empty profiles list. Can empty profiles list cause problems down the path?

j0sh replied:
Good catch. I've been investigating and testing various cases here, and this actually exposes some issues that need to be tackled from multiple angles. Will have some more info and fixes in the next days.

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/551/commits/397944dc9cfb9e18ce4a2ac631e73c3bc49373b2 and https://github.com/livepeer/lpms/pull/89 (vendored LPMS has been updated)

j0sh committed:
eth: Use ethcommon/common in helpers.

j0sh committed:
rpc: Add tests around out-of-range and unclaimable jobs.

j0sh committed:
eth: Add EthJobToDBJob helper.

j0sh committed:
eth: Use db.GetJob during claims recovery.

j0sh commented:
Some more updates, some have been fixed up into existing commits.

* Updated db.GetJob to also return the total number of claims. `job.FirstClaimSubmitted` is calculated by checking if there's at least one claim. Note that this number is a guess at best; the actual number of claims for a given job on the blockchain may differ. This may happen under some unusual transcoder configurations, eg if running multiple nodes. (This is not a new issue; it's been around since day one, but it surfaced while building claims recovery. The claim manager will abort the claim process if the locally inferred claim ID doesn't match what the blockchain eventually gives us.)

* Changed the claimmanager's GetJob to use the local DB. We really don't want lookup failures while trying to do claims.

* Added a ton of tests, including a whole new set of tests around `eth.RecoverClaims`, which is where the claimmanager calls GetJob.

j0sh committed:
vendor: Update LPMS to e867372.

j0sh committed:
db: Error out when GetJob detects an invalid profile.

Also add some other preemptive empty-profile checking.

j0sh commented:
Fixed some issues that could come up with corrupted profiles, including a LPMS crash.

If we somehow receive an empty set of profiles, the goclient will try to claim those even though we don't have any work that needs to be done for them. To preempt the possibility of an accidentally empty set of profiles resulting in slashing, RPC now exits early if a job with an empty profile is found.

This is one area where the user needs to be extremely careful with data tampering, since we aren't reaching out to the blockchain anymore for the ground-truth. Perhaps one day we'll add integrity checks on the DB contents.

yondonfu commented on common/db.go:
```diff
@@ -169,14 +171,16 @@ func NewDBJob(id *big.Int, streamID string,
+		CreationBlock:       big.NewInt(j.startBlock),
+		EndBlock:            big.NewInt(j.endBlock),
+		TotalClaims:         big.NewInt(j.Claims),
+		FirstClaimSubmitted: j.Claims > 0,
```
So, in the current claim-verify workflow, a new claim is inserted into the DB before the node attempts to submit the tx (https://github.com/livepeer/go-livepeer/blob/master/eth/claimmanager.go#L337). Perhaps a slightly more accurate calculation for `FirstClaimSubmitted` (when looking up the job from the DB) could involve:

1. Looking up a claim in the DB with `jobID` and `claimID = 0`. The first on-chain claim should always have an ID of 0.
2. If the claim does not exist in the DB, `FirstClaimSubmitted = false`.
3. Check the claim's status. If the status is "FAIL submit" or "FAIL check tx", `FirstClaimSubmitted = false`. Otherwise, `FirstClaimSubmitted = true`.

j0sh replied:
If we actually proceed with a claim in that situation, the claimmanager would abort it anyway because the ClaimID returned by `db.InsertClaim` doesn't take into account previous failed attempts.

This is definitely something that should be made more robust, but it's a bit more involved since it means fixing up the claim IDs in receipts associated with the failed claims. Or maybe the flow needs to be refactored in some other way. Think we can just make a ticket to track the issue for now?

https://github.com/livepeer/go-livepeer/blob/b26253aa6996b2069722ec658222f230ee81d847/eth/claimmanager.go#L366-L382

j0sh replied:
Modifying the `db.countClaims` query to exclude statuses of "FAIL submit tx" and "FAIL check tx" might help with this (it is used in both `db.GetJob` and `db.InsertClaim`), but I need to look at things a little more closely; we might still need to roll back changes to receipts that failed to have their claims submitted.

https://github.com/livepeer/go-livepeer/blob/b26253aa6996b2069722ec658222f230ee81d847/common/db.go#L307

yondonfu replied:
Ah right. Yeah I agree that we should file a ticket and come back to this problem separately as it is a bit more involved than I originally thought

yondonfu reviewed (approved):
LGTM - potential fixes for handling claim transaction submission failures is being tracked here: https://github.com/livepeer/go-livepeer/issues/560

j0sh commented:
Updated LPMS hash to match LPMS master, no code changes
